### PR TITLE
[SPARK-14525][SQL][FOLLOWUP] Clean up JdbcRelationProvider

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes cleaning up the confusing part in `createRelation` as discussed in https://github.com/apache/spark/pull/12601/files#r80627940

Also, this PR proposes the changes below:
- Add documentation for `batchsize` and `isolationLevel`.
- Move property names into `JDBCOptions` so that they can be managed in a single place. which were, `fetchsize`, `batchsize`, `isolationLevel` and `driver`.
## How was this patch tested?

Existing tests should cover this.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @cloud-fan and @gatorsmile Could you take a look please?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 1]
```diff
@@ -50,67 +51,52 @@ class JdbcRelationProvider extends CreatableRelationProvider
...
-            JdbcUtils.dropTable(conn, table)
-            (true, true)
-          }
-        case (SaveMode.Append, true) => (false, true)
-        case (_, true) => throw new IllegalArgumentException(s"Unexpected SaveMode, '$mode'," +
-          " for handling existing tables.")
-        case (_, false) => (true, true)
-      }
+          case SaveMode.ErrorIfExists =>
+            sys.error(s"Table $table already exists.")
```
Comment (JustinPihony) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: Why are you going back to sys.error? My understanding is that the AnalysisException is more explicit and better. Plus the wording was requested.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 2]
```diff
@@ -50,67 +51,52 @@ class JdbcRelationProvider extends CreatableRelationProvider
...
-        // table_options or partition_options.
-        // E.g., "CREATE TABLE t (name string) ENGINE=InnoDB DEFAULT CHARSET=utf8"
-        val createtblOptions = jdbcOptions.createTableOptions
-        val sql = s"CREATE TABLE $table ($schema) $createtblOptions"
-        val statement = conn.createStatement
-        try {
-          statement.executeUpdate(sql)
-        } finally {
-          statement.close()
+          case SaveMode.Ignore => // Just ignore this case.
```
Comment (JustinPihony) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: This is wrong. It should create & save in this case.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala (hunk 1 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: Yup, I will address this tomorrow. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala (hunk 2 above)
Comment (JustinPihony) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: Also, I'd say add a test if there isn't already one that fails

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: Definitely. I will mark this PR as `[WIP]`. Thanks for your quick look.

**Comment by gatorsmile** [comment]
@HyukjinKwon Will review it after you fix all the issues. 

BTW, I checked the documentation. It sounds like not all the property descriptions of JDBC are accurate. Could you please also update them too? Thanks!

**Comment by HyukjinKwon** [comment]
Sure, will update them together soon!

**Comment by HyukjinKwon** [comment]
@gatorsmile I think it is ready to be reviewed. Please take a look.

**Comment by gatorsmile** [comment]
Regarding the JDBC options, which properties are used for the table reading? Which ones are used for writing? Can you please improve the document and make users easily find the answer?

**Comment by HyukjinKwon** [comment]
Good idea. Yes I will.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 3]
```diff
@@ -50,67 +51,48 @@ class JdbcRelationProvider extends CreatableRelationProvider
...
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: Please explicitly explain the behavior. `// Just ignore this case` is almost useless.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 4]
```diff
@@ -50,67 +51,48 @@ class JdbcRelationProvider extends CreatableRelationProvider
...
       val tableExists = JdbcUtils.tableExists(conn, url, table)
+      if (tableExists) {
+        mode match {
+          case SaveMode.Overwrite =>
+            if (isTruncate && isCascadingTruncateTable(url).contains(false)) {
+              // In this case, we should truncate table and then load.
+              truncateTable(conn, table)
+              saveTable(df, url, table, props)
+            } else {
+              // Otherwise, do not truncate but just drop.
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: This comment looks weird. In RDBMS, users might do `drop and recreate` a table, instead of truncating it. Thus, please change it to

`// Otherwise, do not truncate but just drop and recreate`

**Comment by HyukjinKwon** [comment]
@gatorsmile Would this be sensible if I submit another PR to clean up options for https://github.com/apache/spark/pull/15263#issuecomment-250072008 later as strictly the purpose of this PR is to clean up `JdbcRelationProvider`? 

I'd like to unify all the options within `JDBCOptions` and maybe I could fix the documentation together there if you think it makes sense.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 5]
```diff
@@ -50,67 +51,48 @@ class JdbcRelationProvider extends CreatableRelationProvider
...
-            (true, true)
-          }
-        case (SaveMode.Append, true) => (false, true)
-        case (_, true) => throw new IllegalArgumentException(s"Unexpected SaveMode, '$mode'," +
-          " for handling existing tables.")
-        case (_, false) => (true, true)
-      }
+          case SaveMode.ErrorIfExists =>
+            throw new AnalysisException(
+              s"Table or view '$table' already exists, and SaveMode is set to ErrorIfExists.")
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: `Table or view '$table' already exists. SaveMode: ErrorIfExists.`

File: docs/sql-programming-guide.md [hunk 6]
```diff
@@ -1058,7 +1059,21 @@ the Data Sources API. The following options are supported:
...
+    <td><code>batchsize</code></td>
+    <td>
+      The JDBC batch size, which determines how many rows to insert per round trip. This can help performance on JDBC drivers.
+    </td>
+  </tr>
+
+  <tr>
+    <td><code>isolationLevel</code></td>
+    <td>
+      The transaction isolation level, which applies to current connection. Please refer the documenation in <code>java.sql.Connection</code>.
```
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: `current connection` -> `the current connection`

**Comment by gatorsmile** [comment]
Yeah, @HyukjinKwon . This PR is just to rewrite the implementation. No need to resolve all the comments here.

File: docs/sql-programming-guide.md [hunk 7]
```diff
@@ -1058,7 +1059,21 @@ the Data Sources API. The following options are supported:
       The JDBC fetch size, which determines how many rows to fetch per round trip. This can help performance on JDBC drivers which default to low fetch size (eg. Oracle with 10 rows).
     </td>
   </tr>
-  
+
+  <tr>
+    <td><code>batchsize</code></td>
```
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: Same here. No need to change the document in this PR. How about minimizing the code changes and remove all these unrelated changes? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 8]
```diff
@@ -697,4 +697,27 @@ object JdbcUtils extends Logging {
       getConnection, table, iterator, rddSchema, nullTypes, batchSize, dialect, isolationLevel)
     )
   }
+
+  /**
+   * Creates a table according to the given schema.
+   */
+  def createTable(
+      df: DataFrame,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: why `createTable` takes a `DataFrame` instead of a `StructType`? I think it's clearer to make `schemaString` and this method takes `StrcutType`.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Yeap. I intended to make the changes minimised.  Let me fix them together.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Yeap. I intended to make the changes minimised. Let me fix them together.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 9]
```diff
@@ -697,4 +697,27 @@ object JdbcUtils extends Logging {
...
+   * Creates a table according to the given schema.
+   */
+  def createTable(
+      schema: StructType,
+      url: String,
+      table: String,
+      createTableOptions: String,
+      conn: Connection): Unit = {
+    val strSchema = schemaString(schema, url)
+    // Create the table if the table didn't exist.
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: `didn't` -> `does not`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 10]
```diff
@@ -697,4 +697,27 @@ object JdbcUtils extends Logging {
       getConnection, table, iterator, rddSchema, nullTypes, batchSize, dialect, isolationLevel)
     )
   }
+
+  /**
+   * Creates a table according to the given schema.
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: `Creates a table with a given schema`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 11]
```diff
@@ -19,9 +19,10 @@ package org.apache.spark.sql.execution.datasources.jdbc
 
 import java.util.Properties
 
-import scala.collection.JavaConverters.mapAsJavaMapConverter
+import scala.collection.JavaConverters._
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: revert it back?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: I think both are fine but will fix.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: We should avoid using `x.y.z._`, if possible

**Comment by gatorsmile** [comment]
LGTM pending test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala (hunk 11 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: Yeap, but it seems `import scala.collection.JavaConverters._` is being used a lot across the codebase. So, I thought both are fine.

**Comment by HyukjinKwon** [comment]
@gatorsmile Thank you for a close look!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala (hunk 11 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: This might be an exception. My comment is just a general principle.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: Yup. This is even documented in https://github.com/databricks/scala-style-guide#imports as you might already know. I just thought this is an exceptional case.

**Comment by HyukjinKwon** [comment]
Hi @gatorsmile, would there other things more I should take care of?

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merging to master!

**Comment by yhuai** [comment]
Seems it breaks scala 2.10 compilation. Can you take a look? Thanks!

```
[error] /home/jenkins/workspace/spark-master-compile-sbt-scala-2.10/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala:73: value contains is not a member of Option[Boolean]
[error]             if (isTruncate && isCascadingTruncateTable(url).contains(false)) {
[error]                                                             ^
[error] one error found
[error] (sql/compile:compileIncremental) Compilation failed
[error] Total time: 240 s, completed Oct 7, 2016 11:02:49 AM
Build step 'Execute shell' marked build as failure
Finished: FAILURE
```

**Comment by zsxwing** [comment]
Option.contains is only in Scala 2.11...

**Comment by HyukjinKwon** [comment]
Oh, thank you for pointing this out.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 12]
```diff
@@ -50,67 +51,52 @@ class JdbcRelationProvider extends CreatableRelationProvider
...
     props.putAll(parameters.asJava)
-    val conn = JdbcUtils.createConnectionFactory(url, props)()
 
+    val conn = JdbcUtils.createConnectionFactory(url, props)()
     try {
       val tableExists = JdbcUtils.tableExists(conn, url, table)
+      if (tableExists) {
+        mode match {
+          case SaveMode.Overwrite =>
+            if (isTruncate && isCascadingTruncateTable(url).contains(false)) {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala]: one thing -- please avoid using Option.contains in the future. It is in general actually very confusing and more error prone, especially to people not super familiar with Scala, to use option as a collection (or monad).
